lib/model/content.dart

```diff
@@ -1451,6 +1451,61 @@ class _ZulipContentParser {
     return tableNode ?? UnimplementedBlockContentNode(htmlNode: tableElement);
   }
 
+  List<BlockContentNode> parseMathBlocks(dom.Element element) {
+    assert(_debugParserContext == _ParserContext.block);
+    assert(element.localName == 'p');
+    assert((() {
+      final first = element.nodes.first;
+      return first is dom.Element
+        && first.localName == 'span'
+        && first.className == 'katex-display';
+    })());
+    final debugHtmlNode = kDebugMode ? element : null;
+
+    final blocks = <BlockContentNode>[];
+
+    // The case with the `<br>\n` can happen when at the end of a quote;
+    // it seems like a glitch in the server's Markdown processing,
+    // so hopefully there just aren't any further such glitches.
+    bool hasTrailingBreakNewline = false;
+    if (element.nodes
+          case [..., dom.Element(localName: 'br'), dom.Text(text: '\n')]
+    ) {
+      hasTrailingBreakNewline = true;
+    }
+
+    final length = hasTrailingBreakNewline
+      ? element.nodes.length - 2
+      : element.nodes.length;
+    for (var i = 0; i < length; i++) {
+      final child = element.nodes[i];
+
+      // If there are multiple <span class="katex-display"> nodes in a <p>
+      // each node is interleaved by '\n\n'. Whitespaces are ignored in HTML
+      // on web but each node has `display: block`, which renders each node
+      // on a newline. Since the emitted MathBlockNode are BlockContentNode,
+      // we skip these newlines here to replicate the same behavior as on web.
+      if (child case dom.Text(text: '\n\n')) continue;
+
+      if (child is! dom.Element) {
+        blocks.add(UnimplementedBlockContentNode(htmlNode: child));
+        continue;
+      }
+      if (child.className != 'katex-display') {
+        blocks.add(UnimplementedBlockContentNode(htmlNode: child));
+        continue;
```

Comment:
My performance-paranoia about this code also doesn't like having all these checks for the first child, where we already know the answers to them; particularly since that'll almost always be the only child.

But for this one, there is a clean solution, so let's do that. Before the loop, operate on just the first child; it'll need an `as dom.Element`, but none of the rest of these checks. And it'll duplicate the last few lines of the loop body, but that's OK.

Then the loop runs from 1 to `< length`, so will normally have no iterations at all.

Reply:
Added the suggested optimization for the first child.